File: python/src/threads/agents/transcript.py
"""The forwarded transcript of a handoff (spec/schema/README.md, "Handoff scope"): what the user
and the agent said, and the handing-off turn's tool calls and results, which the target gets as
untrusted reference. The tool lines are capped at the source's spill threshold, the oldest
dropped first; the shared vector is spec/conformance/vectors/handoff-transcripts.json."""

from collections.abc import Sequence

from threads.log import (
    Event,
    ModelResponseEvent,
    ModelResponseRecoveredEvent,
    TextPart,
    ToolCallEvent,
    ToolResultEvent,
    ToolResultLateEvent,
    UserInputEvent,
    WokenEvent,
)
from threads.log.jcs import canonicalize
from threads.result import Ok


def _spoken(event: Event) -> str | None:
    if isinstance(event, UserInputEvent):
        return f"user: {event.data.text}" if isinstance(event.data.text, str) else None
    if isinstance(event, ModelResponseEvent | ModelResponseRecoveredEvent):
        said = "".join(p.text for p in event.data.content if isinstance(p, TextPart))
        return f"assistant: {said}" if said else None
    return None


def _tool_line(event: Event) -> str | None:
    if isinstance(event, ToolCallEvent) and event.data.name != "handoff":
        text = canonicalize(dict(event.data.input))
        if not isinstance(text, Ok):
            raise AssertionError("a recorded input always canonicalizes")
        return f"tool_call {event.data.call_id} {event.data.name}: {text.value}"
    if isinstance(event, ToolResultEvent | ToolResultLateEvent):
        flag = " (error)" if event.data.is_error else ""
        return f"tool_result {event.data.call_id}{flag}: {event.data.preview}"
    return None


def handoff_transcript(events: Sequence[Event], cap_bytes: int) -> str:
    opener = max(
        (i for i, e in enumerate(events) if isinstance(e, UserInputEvent | WokenEvent)),
        default=-1,
    )
    lines: list[str] = []
    tools: list[int] = []
    for i, event in enumerate(events):
        said = _spoken(event)
        tool = _tool_line(event) if i > opener else None
        if said is not None:
            lines.append(said)
        elif tool is not None:
            tools.append(len(lines))
            lines.append(tool)
    total = sum(len(lines[i].encode()) for i in tools) + max(len(tools) - 1, 0)
    dropped: list[int] = []
    while tools and total > cap_bytes:
        first = tools.pop(0)
        total -= len(lines[first].encode()) + (1 if tools else 0)
        dropped.append(first)
    if not dropped:
        return "\n".join(lines)
    marker = f"[{len(dropped)} earlier tool lines dropped]"
    gone = set(dropped[1:])
    return "\n".join(
        marker if i == dropped[0] else line for i, line in enumerate(lines) if i not in gone
    )

File: typescript/packages/core/src/loop/agents/transcript.ts
import { responseText } from "../../fold/state";
import type { KnownEvent } from "../../log";
import { canonicalize } from "../../log/jcs";

// The forwarded transcript of a handoff (spec/schema/README.md, "Handoff scope"): what the user
// and the agent said, and the handing-off turn's tool calls and results, which the target gets
// as untrusted reference. The tool lines are capped at the source's spill threshold, the oldest
// dropped first; the shared vector is spec/conformance/vectors/handoff-transcripts.json.

const encoder = new TextEncoder();
const size = (line: string): number => encoder.encode(line).length;

function toolLine(e: KnownEvent): string | undefined {
  if (e.type === "tool_call" && e.data.name !== "handoff") {
    const input = canonicalize(e.data.input);
    if (!input.ok) throw new Error("a recorded input always canonicalizes");
    return `tool_call ${e.data.call_id} ${e.data.name}: ${input.value}`;
  }
  if (e.type === "tool_result" || e.type === "tool_result_late") {
    const flag = e.data.is_error ? " (error)" : "";
    return `tool_result ${e.data.call_id}${flag}: ${e.data.preview}`;
  }
  return undefined;
}

function spoken(e: KnownEvent): string | undefined {
  if (e.type === "user_input")
    return e.data.text === undefined ? undefined : `user: ${e.data.text}`;
  if (e.type !== "model_response" && e.type !== "model_response_recovered")
    return undefined;
  const said = responseText(e.data.content);
  return said === "" ? undefined : `assistant: ${said}`;
}

export function handoffTranscript(
  events: readonly KnownEvent[],
  capBytes: number,
): string {
  const opener = events.findLastIndex(
    (e) => e.type === "user_input" || e.type === "woken",
  );
  const lines: string[] = [];
  const tools: number[] = [];
  for (const [i, e] of events.entries()) {
    const said = spoken(e);
    const tool = i > opener ? toolLine(e) : undefined;
    if (said !== undefined) lines.push(said);
    else if (tool !== undefined) {
      tools.push(lines.length);
      lines.push(tool);
    }
  }
  let total =
    tools.reduce((n, i) => n + size(lines[i] ?? ""), 0) +
    Math.max(tools.length - 1, 0);
  const dropped: number[] = [];
  while (total > capBytes) {
    const first = tools.shift();
    if (first === undefined) break;
    total -= size(lines[first] ?? "") + (tools.length > 0 ? 1 : 0);
    dropped.push(first);
  }
  const [marker, ...gone] = dropped;
  return lines
    .flatMap((line, i) =>
      i === marker
        ? [`[${dropped.length} earlier tool lines dropped]`]
        : gone.includes(i)
          ? []
          : [line],
    )
    .join("\n");
}
